File: flink/src/main/java/nl/cowbird/flinkbenchmark/MessageStream.java
package nl.cowbird.flinkbenchmark;

import java.util.ArrayList;

/**
 * Created by gdibernardo on 30/05/2017.
 */

public class MessageStream {

    private Boolean readyForReduction;


    private ArrayList<Message> stream;


    private Double minimumElementInStream = Double.MAX_VALUE;

    private Double maximumElementInStream = Double.MIN_VALUE;

    private Double sum = 0.0;


    private Long firstTimestamp = Long.MAX_VALUE;


    public MessageStream() {
        stream = new ArrayList<>();
        readyForReduction = false;
    }


    public int getStreamCount() {
        return stream.size();
    }


    public Boolean isReadyForReduction() {
        return readyForReduction;
    }


    public void setReadyForReduction(Boolean value) {
        readyForReduction = value;
    }


    public void addMessage(Message message) {

        if(message.payload > maximumElementInStream)
            maximumElementInStream = message.payload;

        if(message.payload < minimumElementInStream)
            minimumElementInStream = message.payload;

        if(message.emittedAt < firstTimestamp)
            firstTimestamp = message.emittedAt;

        sum += message.payload;

        stream.add(message);
    }


    public String messageId() {
        return stream.get(0).id;
    }


    public String reductionOperator() {
        return stream.get(0).reduction;
    }


    public Long firsIngestionTime() {
        return firstTimestamp;
    }


    public Double getMin() {
        return minimumElementInStream;
    }


    public Double getMax() {
        return maximumElementInStream;
    }


    public Double getSum() {
        return sum;
    }
}

File: spark-streaming/sparkbenchmark-kafka010/src/main/scala/nl/cowbird/sparkbenchmark/MessageStream.scala
package nl.cowbird.sparkbenchmark

/**
  * Created by gdibernardo on 31/05/2017.
  */
class MessageStream extends Serializable {

  private var firstTimestamp: Long = Long.MaxValue

  private var minimumValue: Double = Double.MaxValue
  private var maximumValue: Double = Double.MinValue
  private var sumValues: Double = 0

  private var stream: Seq[Message] = Seq[Message]()

  var readyForReduction: Boolean = false

  def addMessage(message: Message): Unit = {
    if(message.payload > maximumValue) {
      maximumValue = message.payload
    }
    if(message.payload < minimumValue) {
      minimumValue = message.payload
    }

    if(message.emittedAt < firstTimestamp) {
      firstTimestamp = message.emittedAt
    }

    sumValues += message.payload

    stream = stream :+ message
  }


  def streamCount(): Int = {
    return stream.length
  }


  def firstIngestionTime(): Long = {
    return firstTimestamp
  }


  def min(): Double = {
    return minimumValue
  }


  def max(): Double = {
    return maximumValue
  }


  def sum(): Double = {
    return sumValues
  }

  def reductionOperator(): String = {
    return stream(0).reduction
  }


  def messageId(): String = {
    return stream(0).id
  }

  def getStream(): Seq[Message] = {
    return stream
  }
}
